include/openthread/srp_client.h

```diff
@@ -0,0 +1,489 @@
+ * `OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL` and the related ones for more details).
+ *
+ *
+ * @param[in] aContext          A pointer to an arbitrary context (provided when callback was registered).
+ * @param[in] aError            The error (see above).
+ * @param[in] aHostInfo         A pointer to host info.
+ * @param[in] aService          The head of linked-list containing all services (excluding the ones removed). NULL if
+ *                              the list is empty.
+ * @param[in] aRemovedServices  The head of linked-list containing all removed services. NULL if the list is empty.
+ *
+ */
+typedef void (*otSrpClientCallback)(void *                     aContext,
+                                    otError                    aError,
+                                    const otSrpClientHostInfo *aHostInfo,
+                                    const otSrpClientService * aServices,
+                                    const otSrpClientService * aRemovedServices);
+
+/**
+ * This function starts the SRP client operation.
+ *
+ * SRP client will prepare and send "SRP Update" message to the SRP server once all the following conditions are met:
+ *
+ *  - The SRP client is started - `otSrpClientStart()` is called.
+ *  - Host name is set - `otSrpClientSetHostName()` is called.
+ *  - At least one host IPv6 address is set - `otSrpClientSetHostName()` is called.
+ *  - At least one service is added - `otSrpClientAddService()` is called.
+ *
+ * It does not matter in which order these functions are called. When all conditions are met, the SRP client will
+ * wait for a short delay before preparing an "SRP Update" message and sending it to server. This delay allows for user
+ * to add multiple services and/or IPv6 addresses before the first SRP Update message is sent (ensuring a single SRP
+ * Update is sent containing all the info).
+ *
+ * @param[in] aInstance        A pointer to the OpenThread instance.
+ * @param[in] aServerSockAddr  The socket address (IPv6 address and port number) of the SRP server.
+ * @param[in] aCallback        The callback which is used to notify events and changes. Can be NULL if not needed.
+ * @param[in] aContext         An arbitrary context used with @p aCallback.
+ *
+ * @retval OT_ERROR_NONE       SRP client operation started successfully.
+ * @retval OT_ERROR_ALREADY    SRP client is already started and running.
+ *
+ */
+otError otSrpClientStart(otInstance *        aInstance,
+                         const otSockAddr *  aServerSockAddr,
+                         otSrpClientCallback aCallback,
+                         void *              aContext);
+
+/**
+ * This function stops the SRP client operation.
+ *
+ * This function stops any further interactions with the SRP server. Note that it does not remove or clear host info
+ * and/or list of services. It marks all services to be added/removed again once the client is started again.
```

Comment:
The idea is that when a client is stopped and restarted again any previously registered services would register again (as if they are being added again). Or if we are in middle of removing a service, we mark it to be removed once client is re-started. The reason that we need this, is that the user may `Start` with a possibly new SRP server.

We also have `Pause` mechanism which we use internally when device is disconnected/offline. For now, we do not expose the `Pause` as an external API since I don't think there is any immediate use for it (we can add later if there is a use for it). 

The main difference between `Stop()` and `Pause()` is that on `Pause` we keep the previously registered services as they are (i.e. we won't re-add/re-register them on `Resume`).